Regression with fast

I can't compile this code anymore

```
import numpy as np

#pythran export xxx(int32[][] list, int32[][])
def xxx(a, b=None):
    if b is None:
        assert len(a) > 0
        b = np.copy(a[0])
        a = a[1:]
    else:
        b = np.copy(b)
    m = b >= 0
    for array in a:
        b[m] *= array[m]
return b
```

> /home/yann/virtualenvs/pythran/lib/python3.5/site-packages/pythran-0.8.6.dev0+x-py3.5.egg/pythran/pythonic/types/numpy_texpr.hpp:171:12: error: no viable conversion from returned value of type 'typename std::enable_if<is_numexpr_arg<ndarray<long, pshape<long> > >::value && !std::is_same<bool, typename ndarray<long, pshape<long> >::dtype>::value, ndarray<dtype, pshape<long, long> > >::type' (aka '(anonymous namespace)::pythonic::types::ndarray<int, (anonymous namespace)::pythonic::types::pshape<long, long> >') to function return type 'typename std::enable_if<is_numexpr_arg<ndarray<bool, pshape<long, long> > >::value && std::is_same<bool, typename ndarray<bool, pshape<long, long> >::dtype>::value, numpy_vexpr<numpy_texpr_2<ndarray<int, pshape<long, long> > >, ndarray<long, pshape<long> > > >::type' (aka '(anonymous namespace)::pythonic::types::numpy_vexpr<(anonymous namespace)::pythonic::types::numpy_texpr_2<(anonymous namespace)::pythonic::types::ndarray<int, (anonymous namespace)::pythonic::types::pshape<long, long> > >, (anonymous namespace)::pythonic::types::ndarray<long, (anonymous namespace)::pythonic::types::pshape<long> > >')
>     return this->fast(
>

--- a/pythran/pythonic/include/types/numpy_texpr.hpp
+++ b/pythran/pythonic/include/types/numpy_texpr.hpp
@@ -95,9 +95,10 @@ namespace types
     fast(F const &filter) const;
 
     template <class F> // indexing through an array of indices -- a view
-    typename std::enable_if<is_numexpr_arg<F>::value &&
-                                !std::is_same<bool, typename F::dtype>::value,
-                            ndarray<dtype, pshape<long, long>>>::type
+    typename std::enable_if<
+        is_numexpr_arg<F>::value &&
+            !std::is_same<bool, typename F::dtype>::value,
+        numpy_vexpr<numpy_texpr_2, ndarray<long, pshape<long>>>>::type
     fast(F const &filter) const;
 
     template <class F> // indexing through an array of boolean -- a mask
@@ -108,9 +109,10 @@ namespace types
     operator[](F const &filter) const;
 
     template <class F> // indexing through an array of indices -- a view
-    typename std::enable_if<is_numexpr_arg<F>::value &&
-                                !std::is_same<bool, typename F::dtype>::value,
-                            ndarray<dtype, pshape<long, long>>>::type
+    typename std::enable_if<
+        is_numexpr_arg<F>::value &&
+            !std::is_same<bool, typename F::dtype>::value,
+        numpy_vexpr<numpy_texpr_2, ndarray<long, pshape<long>>>>::type
     operator[](F const &filter) const;
     auto operator[](long i) const -> decltype(this->fast(i));
     auto operator[](long i) -> decltype(this->fast(i));

--- a/pythran/pythonic/types/ndarray.hpp
+++ b/pythran/pythonic/types/ndarray.hpp
@@ -751,9 +751,8 @@ namespace types
       if (filter.fast(i))
         raw[n++] = i;
     // realloc(raw, n * sizeof(long));
-    long shp[1] = {n};
-    return this->fast(
-        ndarray<long, pshape<long>>(raw, shp, types::ownership::owned));
+    return this->fast(ndarray<long, pshape<long>>(raw, pshape<long>(n),
+                                                  types::ownership::owned));
   }
 
   template <class T, class pS>

--- a/pythran/pythonic/types/numpy_texpr.hpp
+++ b/pythran/pythonic/types/numpy_texpr.hpp
@@ -167,9 +167,8 @@ namespace types
       if (filter.fast(i))
         raw[n++] = i;
     // realloc(raw, n * sizeof(long));
-    long shp[1] = {n};
-    return this->fast(
-        ndarray<long, pshape<long>>(raw, shp, types::ownership::owned));
+    return this->fast(ndarray<long, pshape<long>>(raw, pshape<long>(n),
+                                                  types::ownership::owned));
   }
 
   template <class E>
@@ -187,37 +186,26 @@ namespace types
   template <class F> // indexing through an array of indices -- a view
   typename std::enable_if<
       is_numexpr_arg<F>::value && !std::is_same<bool, typename F::dtype>::value,
-      ndarray<typename numpy_texpr_2<E>::dtype, pshape<long, long>>>::type
+      numpy_vexpr<numpy_texpr_2<E>, ndarray<long, pshape<long>>>>::type
       numpy_texpr_2<E>::
       operator[](F const &filter) const
   {
     static_assert(F::value == 1,
                   "advanced indexing only supporint with 1D index");
 
-    ndarray<typename numpy_texpr_2<E>::dtype, pshape<long, long>> out(
-        array<long, 2>{{filter.flat_size(), std::get<1>(shape())}},
-        none_type());
-    std::transform(
-        filter.begin(), filter.end(), out.begin(),
-        [this](typename F::dtype index) { return operator[](index); });
-    return out;
+    return {*this, filter};
   }
 
   template <class E>
   template <class F> // indexing through an array of indices -- a view
   typename std::enable_if<
       is_numexpr_arg<F>::value && !std::is_same<bool, typename F::dtype>::value,
-      ndarray<typename numpy_texpr_2<E>::dtype, pshape<long, long>>>::type
+      numpy_vexpr<numpy_texpr_2<E>, ndarray<long, pshape<long>>>>::type
   numpy_texpr_2<E>::fast(F const &filter) const
   {
     static_assert(F::value == 1,
                   "advanced indexing only supporint with 1D index");
-    ndarray<typename numpy_texpr_2<E>::dtype, pshape<long, long>> out(
-        array<long, 2>{{filter.flat_size(), std::get<1>(shape())}},
-        none_type());
-    std::transform(filter.begin(), filter.end(), out.begin(),
-                   [this](typename F::dtype index) { return fast(index); });
-    return out;
+    return {*this, filter};
   }
 
   template <class E>

--- a/pythran/tests/test_ndarray.py
+++ b/pythran/tests/test_ndarray.py
@@ -871,3 +871,22 @@ def assign_ndarray(t):
         self.run_test(code, 10, test_fexpr1=[int])
 
 
+    def test_vexpr0(self):
+        code = '''
+            import numpy as np
+            def vexpr0(a, b=None):
+                if b is None:
+                    assert len(a) > 0
+                    b = np.copy(a[0])
+                    a = a[1:]
+                else:
+                    b = np.copy(b)
+                m = b >= 0
+                for array in a:
+                    b[m] *= array[m]
+                return b'''
+        self.run_test(code,
+            [numpy.arange(10, dtype=numpy.int32).reshape(5,2)],
+            2 * numpy.arange(10, dtype=numpy.int32).reshape(5,2),
+            vexpr0=[List[NDArray[numpy.int32,:,:]], NDArray[numpy.int32,:,:]])
+

